File: daily/30 -day finish 300 coding challenge/2020-5-14. 10 questions - Tree/173. Binary Search Tree Iterator.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BSTIterator:
    def __init__(self, root):
        # Flatten binary search tree
        self.stack = []
        curr = root
        while curr:
            self.stack.append(curr)
            curr = curr.left
    
    def next(self):
        if self.stack:
            node = self.stack.pop()
        if node.right:
            curr = node.right
            while curr:
                self.stack.append(curr)
                curr = curr.left
        return node.val
    
    def hasNext(self):
        return len(self.stack) > 0

# time complexity o(N)
# space complexity o(N)

File: Leetcode First 200 JavaScript solution/173. Binary Search Tree Iterator.js
var BSTIterator = function (root) {
    stack = [];
    while (root) {
        stack.unshift(root);
        root = root.left;
    }
};

BSTIterator.prototype.next = function() {
    root = stack[0];
    stack.shift();
    value = root.val;
    root = root.right;
    while (root) {
        stack.unshift(root);
        root = root.left;
    }
    return value;
}

BSTIterator.prototype.hasNext = function() {
    return stack.length;
}

// time complexity o(n)
// space complexity o(n)

